### ISSUE TITLE
Error 404 when publishing

### ISSUE THREAD

**Issue by MarkTNO** [GeoCat/qgis-bridge-plugin#141]
When trying to publish polygon data I get the errors below:
Using QGIS 3.16.16 and geocat bridge 4.1.3
Any help is greatly appreciated!

Warnings: 
Errors:
Failed to create new style 'aw_vergunningen_all' in workspace 'testGeoserver' using ZIP file 'C:\Users\vrijlandtmaw\AppData\Local\Temp\geocatbridge\94391cb763f848c8ac8f9ed363134a26\aw_vergunningen_all.zip': 400 Client Error: for url: http://localhost:6480/geoserver/rest/workspaces/testGeoserver/styles?name=aw_vergunningen_all
Traceback (most recent call last): 
File "C:/Users/vrijlandtmaw/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\geocatbridge\publish\tasks.py", line 146, in run 
publishLayer(layer, safe_name) 
File "C:/Users/vrijlandtmaw/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\geocatbridge\publish\tasks.py", line 94, in publishLayer 
self.geodata_server.publishLayer(lyr, fields) 
File "C:/Users/vrijlandtmaw/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\geocatbridge\servers\models\geoserver.py", line 237, in publishLayer 
self._publishVectorLayerFromFileToPostgis(layer, filename) 
File "C:/Users/vrijlandtmaw/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\geocatbridge\servers\models\geoserver.py", line 441, in _publishVectorLayerFromFileToPostgis 
ret = self.request(url, "post", body) 
File "C:/Users/vrijlandtmaw/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\geocatbridge\servers\bases.py", line 225, in request 
result.raise_for_status() 
File "C:\PROGRA~1\QGIS3~1.10\apps\Python37\lib\site-packages\requests\models.py", line 941, in raise_for_status 
raise HTTPError(http_error_msg, response=self) 
requests.exceptions.HTTPError: 404 Client Error: for url: http://localhost:6480/geoserver/rest/imports.json

**Comment by GeoSander** [comment]
Hi @MarkTNO,

From the traceback, it seems that the URL http://localhost:6480/geoserver/rest/imports.json does not exist. This leads me to believe that you may not have installed the [GeoServer Importer extension](https://docs.geoserver.org/stable/en/user/extensions/importer/index.html), which is required if you are publishing data to GeoServer using the REST API (as mentioned in the [Bridge docs](https://www.geocat.net/docs/bridge/qgis/latest/installation.html?highlight=importer#geoserver)).

You can perform a quick check by navigating to http://localhost:6480/geoserver/rest/about/status.json in your case. This should return a JSON document with some status objects. There should be at least one object in there with `"name": "Importer REST"` and one with `"name": "Importer Core Extension"` or something similar.

Let me know how it goes.

Sander

**Comment by MarkTNO** [comment]
Hi @GeoSander,

Thanks for the quick reply!
I updated docker compose see below (STABLE_EXTENSIONS) and now I manage to get a bit further.
>  geoserver:
    container_name: esa-geoserver
    image: kartoza/geoserver:${GEOSERVER_VERSION_TAG}
    volumes:
      - geoserver_storage:/opt/geoserver/data_dir
    ports:
      - "${ESA_GEOSERVER_PORT:-8080}:8080"
    restart: on-failure
    environment:
      - GEOSERVER_DATA_DIR=${GEOSERVER_DATA_DIR}
      - GEOWEBCACHE_CACHE_DIR=${GEOWEBCACHE_CACHE_DIR}
      - GEOSERVER_ADMIN_PASSWORD=${GEOSERVER_ADMIN_PASSWORD}
      - GEOSERVER_ADMIN_USER=${GEOSERVER_ADMIN_USER}
      - INITIAL_MEMORY=${INITIAL_MEMORY}
      - MAXIMUM_MEMORY=${MAXIMUM_MEMORY}
      - STABLE_EXTENSIONS=${IMPORTER_PLUGIN}
    healthcheck:
      test: curl --fail -s http://localhost:8080/ || exit 1
      interval: 1m30s
      timeout: 10s
      retries: 3
    deploy:
      placement:
        constraints: [node.labels.postgres == true]



But now I get the following error (this number '4' at the end keep increasing with each try):

Warnings: 
Errors:
Failed to create new style 'aw_vergunningen_all' in workspace 'testGeoserver' using ZIP file 'C:\Users\vrijlandtmaw\AppData\Local\Temp\geocatbridge\02e62913ea6d439087b0eeaf6dac20cc\aw_vergunningen_all.zip': 400 Client Error: for url: http://localhost:6480/geoserver/rest/workspaces/testGeoserver/styles?name=aw_vergunningen_all
Failed to publish QGIS layer 'AW_Vergunningen_ALL' as 'aw_vergunningen_all'. 

GeoServer Importer Extension error: Schema 'http://testGeoserver:aw_vergunningen_all4' does not exist.

**Comment by GeoSander** [comment]
Hi @MarkTNO,

Good to hear that you got a bit further.

> this number '4' at the end keep increasing with each try

Yes, this is due to the fact that GeoServer wants unique layer names across all workspaces.
Normally, when you republish the same layer again to the same workspace, Bridge will first erase the entire workspace. This is by design: we are working on a new version though that will allow you to add stuff to an existing workspace without erasing the entire workspace. Once the workspace is erased, the layer can be published using the original name (without the numeric prefix). However, there is one caveat: if you publish to GeoServer using the REST API (the ["handled by GeoServer"](https://www.geocat.net/docs/bridge/qgis/latest/server_configuration.html#import-into-postgis-database-handled-by-geoserver) option and you're using PostGIS (seems you are?), the database tables for the layers will **not** be deleted. The GeoServer REST API that Bridge is using does not have the means to do so (probably a security matter). This means that if you keep on republishing `aw_vergunningen_all`, the underlying data table for it may exist several times in PostGIS, but only the one with the highest numeric suffix is actually linked to the layer and the other ones are orphans. This is sloppy, but there is no other way...

However, since you are running everything locally, you could also [let Bridge manage PostGIS directly](https://www.geocat.net/docs/bridge/qgis/latest/server_configuration.html#import-into-postgis-database-direct-connect), which *should* result in a cleaner workflow and you also don't need the Importer extension actually.

First of all though, I would clean up your GeoServer workspaces, because things may have gotten a bit messy due to the missing Importer extension and Bridge partially publishing things. So please go to the GeoServer admin web page and remove all layers, styles, and workspaces related to `testGeoserver` and/or `aw_vergunningen_all*`. Also, open a DB client and remove all tables related to `aw_vergunningen_all*` that you can find in your PostGIS DB. Then, shut down the container(s) and pull them back up again, just to be sure.

**Comment by MarkTNO** [comment]
Hi @GeoSander,

It turns out it is not so easy to use raster data with postgis in Geoserver (https://docs.geoserver.org/latest/en/user/data/raster/postgisraster.html, but I can't find this JDBC Image Mosaic extension on geoserver.org).
Now I am using 'File-based storage' which allows me to upload all types of data (it creates a separate store per layer, is that ok?). However I'm not able to publish the symbology:
![image](https://user-images.githubusercontent.com/25008935/151559554-b828a9fd-4e7e-40e9-9696-2acc99c913b8.png)

I tried clearing all the geoserver layers and restarting the containers.

Kind regards,
Mark

**Comment by GeoSander** [comment]
Hi @MarkTNO,

> it creates a separate store per layer, is that ok?
No, it's not, but that is how Bridge works (I am changing that behavior at the moment)...

However, that 400 should not occur. Could you please check the GeoServer logs? You can attach a zipped log here, if you like (log in as admin in the Web UI and go to "GeoServer Logs" in the menu and then click the "Download the full log file" link at the bottom).

Also, if the error above pops up, leave Bridge and QGIS open and navigate to the temporary directory containing the zipped SLD as mentioned in the error message. Please attach that zip file here too.

Sander

**Comment by MarkTNO** [comment]
Hi @GeoSander,

The online GeoServer Logs remains empty. I tried having it open while publishing and refreshing after but it still remains empty. Also 'Download the full log file' gave an empty file.

I tried publishing two layers: a raster file and a shape file. Both files appear as layers on GeoServer, but without style file. The colors for the raster are published (since they are part of the tif file?).

Locally .sld files are being created for both files (see below for the shapefile) and a .gpkg file for the shapefile and a .tif file for the raster file.

Groeten,
Mark

![image](https://user-images.githubusercontent.com/25008935/152333508-abc0f05e-e2b6-4b37-b1d3-41fb1a3c25ad.png)


Below is the QGIS GeoCat Bridge log: 
`2022-02-03T12:18:42     INFO    GET http://localhost:6480/geoserver/rest/about/version
2022-02-03T12:18:49     INFO    GET http://localhost:6480/geoserver/rest/workspaces.json
2022-02-03T12:18:49     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/layers.json
2022-02-03T12:18:50     INFO    GET http://localhost:6480/geoserver/rest/about/version.json
2022-02-03T12:18:50     INFO    GET http://localhost:6480/geoserver/rest/workspaces.json
2022-02-03T12:18:50     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/datastores.json
2022-02-03T12:18:50     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/datastores/aw_vergunningen_all.json
2022-02-03T12:18:50     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver.json
2022-02-03T12:18:50     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/datastores/mar-2020_nlog_fields_utm.json
2022-02-03T12:18:50     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver.json
2022-02-03T12:18:50     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/datastores/rosl_roslu__ntg_points.json
2022-02-03T12:18:50     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver.json
2022-02-03T12:18:50     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/datastores/testGeoserver.json
2022-02-03T12:18:51     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver.json
2022-02-03T12:18:51     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/styles.json
2022-02-03T12:18:51     INFO    DELETE http://localhost:6480/geoserver/rest/workspaces/testGeoserver.json?recurse=true
2022-02-03T12:18:51     INFO    POST http://localhost:6480/geoserver/rest/workspaces
2022-02-03T12:18:51     INFO    POST http://localhost:6480/geoserver/rest/workspaces/testGeoserver/datastores.json
2022-02-03T12:18:51     INFO    POST http://localhost:6480/geoserver/rest/workspaces/testGeoserver/datastores.json
2022-02-03T12:18:51     INFO    POST http://localhost:6480/geoserver/rest/workspaces/testGeoserver/datastores.json
2022-02-03T12:18:51     INFO    POST http://localhost:6480/geoserver/rest/workspaces/testGeoserver/datastores.json
2022-02-03T12:18:51     INFO    GET http://localhost:6480/geoserver/rest/workspaces.json
2022-02-03T12:18:52     INFO    Style for layer 'AW_Vergunningen_ALL' exported as ZIP file to 'C:\Users\vrijlandtmaw\AppData\Local\Temp\geocatbridge\ab18054001f54327bccc046289bd4dd1\aw_vergunningen_all.zip'
2022-02-03T12:18:52     INFO    GET http://localhost:6480/geoserver/rest/workspaces.json
2022-02-03T12:18:52     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/styles.json
2022-02-03T12:18:52     INFO    POST http://localhost:6480/geoserver/rest/workspaces/testGeoserver/styles?name=aw_vergunningen_all
2022-02-03T12:18:52     CRITICAL    Failed to create new style 'aw_vergunningen_all' in workspace 'testGeoserver' using ZIP file 'C:\Users\vrijlandtmaw\AppData\Local\Temp\geocatbridge\ab18054001f54327bccc046289bd4dd1\aw_vergunningen_all.zip': 400 Client Error: for url: http://localhost:6480/geoserver/rest/workspaces/testGeoserver/styles?name=aw_vergunningen_all
2022-02-03T12:18:52     INFO    Layer AW_Vergunningen_ALL exported to C:\Users\vrijlandtmaw\AppData\Local\Temp\geocatbridge\bd5744b897e74d629777e43e0fb3da69\aw_vergunningen_all.gpkg
2022-02-03T12:18:52     INFO    Publishing layer from file: C:\Users\vrijlandtmaw\AppData\Local\Temp\geocatbridge\bd5744b897e74d629777e43e0fb3da69\aw_vergunningen_all.gpkg
2022-02-03T12:18:52     INFO    DELETE http://localhost:6480/geoserver/rest/workspaces/testGeoserver/datastores/aw_vergunningen_all?recurse=true
2022-02-03T12:18:52     INFO    PUT http://localhost:6480/geoserver/rest/workspaces/testGeoserver/datastores/aw_vergunningen_all/file.gpkg?update=overwrite
2022-02-03T12:18:52     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/datastores/aw_vergunningen_all/featuretypes/aw_vergunningen_all.json
2022-02-03T12:18:52     INFO    PUT http://localhost:6480/geoserver/rest/workspaces/testGeoserver/datastores/aw_vergunningen_all/featuretypes/aw_vergunningen_all.json
2022-02-03T12:18:52     INFO    Successfully created feature type from GeoPackage file 'C:\Users\vrijlandtmaw\AppData\Local\Temp\geocatbridge\bd5744b897e74d629777e43e0fb3da69\aw_vergunningen_all.gpkg'
2022-02-03T12:18:52     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/layers/aw_vergunningen_all.json
2022-02-03T12:18:52     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/styles.json
2022-02-03T12:18:52     WARNING    Style 'aw_vergunningen_all' does not exist in workspace 'testGeoserver'
2022-02-03T12:18:53     INFO    Style for layer 'NLFFS_top' exported as ZIP file to 'C:\Users\vrijlandtmaw\AppData\Local\Temp\geocatbridge\d53b36e4b33041878c9e0cf3c478520b\nlffs_top.zip'
2022-02-03T12:18:53     INFO    GET http://localhost:6480/geoserver/rest/workspaces.json
2022-02-03T12:18:53     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/styles.json
2022-02-03T12:18:53     INFO    POST http://localhost:6480/geoserver/rest/workspaces/testGeoserver/styles?name=nlffs_top
2022-02-03T12:18:53     CRITICAL    Failed to create new style 'nlffs_top' in workspace 'testGeoserver' using ZIP file 'C:\Users\vrijlandtmaw\AppData\Local\Temp\geocatbridge\d53b36e4b33041878c9e0cf3c478520b\nlffs_top.zip': 400 Client Error: for url: http://localhost:6480/geoserver/rest/workspaces/testGeoserver/styles?name=nlffs_top
2022-02-03T12:18:54     INFO    Layer NLFFS_top exported to C:\Users\vrijlandtmaw\AppData\Local\Temp\geocatbridge\34ac34b8e6f948f6b896d01e711a69df\nlffs_top.tif
2022-02-03T12:18:54     INFO    GET http://localhost:6480/geoserver/rest/workspaces.json
2022-02-03T12:18:54     INFO    PUT http://localhost:6480/geoserver/rest/workspaces/testGeoserver/coveragestores/nlffs_top/file.geotiff
2022-02-03T12:18:55     INFO    Successfully created coverage from TIFF file 'C:\Users\vrijlandtmaw\AppData\Local\Temp\geocatbridge\34ac34b8e6f948f6b896d01e711a69df\nlffs_top.tif'
2022-02-03T12:18:55     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/layers/nlffs_top.json
2022-02-03T12:18:55     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/styles.json
2022-02-03T12:18:55     WARNING    Style 'nlffs_top' does not exist in workspace 'testGeoserver'
2022-02-03T12:18:55     INFO    DELETE http://localhost:6480/geoserver/rest/workspaces/testGeoserver/layergroups/testgroup
2022-02-03T12:18:55     CRITICAL    Could not create layer groups: HTTPConnectionPool(host='localhost', port=6480): Max retries exceeded with url: /geoserver/rest/workspaces/testGeoserver/layergroups/testgroup (Caused by ResponseError('too many 500 error responses'))
2022-02-03T12:19:24     INFO    GET http://localhost:6480/geoserver/rest/about/version
2022-02-03T12:19:24     INFO    GET http://localhost:6480/geoserver/rest/workspaces/testGeoserver/layers.json`

**Comment by GeoSander** [comment]
Hi @MarkTNO,

Thanks for the info. I have removed the Bridge log again though, as it didn't tell us anything new.

Since you're using a pre-configured Docker image, I assume that logging has been set to only output critical errors or something. Please go to Settings > Global on the GeoServer admin page and choose the `DEFAULT_LOGGING.properties` or even `VERBOSE_LOGGING.properties` profile:

![image](https://user-images.githubusercontent.com/8972335/152437872-c47604db-6e40-4e21-aee2-1c4322957d41.png)

It could also be that - because you're on Docker - it's trying to write the log somewhere and it doesn't have permissions. Did you map a Docker volume for the GeoServer data directory? Then it should write a log file in the `logs` folder in there, by default. So the file may be on your system already.

Either way, **please attach the SLD here**, because that may tell me what's wrong with it.
Actually, what happens when you manually upload that SLD using the GeoServer admin panel? Go to Data > Styles > +Add a new style and then press the `Choose File` button:

![image](https://user-images.githubusercontent.com/8972335/152439119-f607bd41-ea6a-4620-aa37-8d366115aa0c.png)

**Comment by MarkTNO** [comment]
Hi @GeoSander,

I restarted the container and tried to publish a shapefile.
The Geoserver Logs are attached below.
[geoserver.log](https://github.com/GeoCat/qgis-bridge-plugin/files/8001009/geoserver.log)
And the sld file is here:
[aw_vergunningen_all.zip](https://github.com/GeoCat/qgis-bridge-plugin/files/8001011/aw_vergunningen_all.zip)

I tried to add the style manually as you suggested. I could create a new style with this sld file, 'Validate' did not give any errors but when I apply that style to the layer it does not work. When I changed encoding of the xml from "utf8" to "UTF-8" it did work. That might be the cause of the geocat bridge problem.
![image](https://user-images.githubusercontent.com/25008935/152498400-e22c4c13-dce5-4703-b522-15f00cefc61f.png)
